Add support for `org.apache.spark.sql.catalyst.expressions.Bin`

ustcfy opened the pull request:
<!--

Thank you for contributing to RAPIDS Accelerator for Apache Spark!

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

4. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

5. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present).

6. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please avoid
   rebasing your branch during the review process, as this causes the context
   of any comments made by reviewers to be lost. If conflicts occur during
   review then they should be resolved by merging into the branch used for
   making the pull request.

Many thanks in advance for your cooperation!

-->
Closes https://github.com/NVIDIA/spark-rapids/issues/11648

Based on https://github.com/NVIDIA/spark-rapids-jni/pull/2760, this PR adds support for `org.apache.spark.sql.catalyst.expressions.Bin`.

Thanks for the guidance provided by @revans2 !

I ran
```
spark.time(spark.range(0, 2000000000L, 1, 32).selectExpr("sum(octet_length(bin(id)))").show())
```
and
```
spark.time(spark.range(0, 2000000000L, 1, 32).selectExpr("sum(octet_length(id))").show())
```
separately to calculate the time difference between the two. 

The results showed that the `GpuBin` (TITAN RTX) demonstrated approximately 5 times speedup compared to the `CpuBin` (8 CPU cores).

revans2 commented:
The GPU is consistently slower than the CPU is at this. Have you done any profiling so we can understand what is happening?

revans2 reviewed (commented):
The code here looks fine. I just want to understand why we are losing to the CPU and if we can fix that before we check it in, or if we should file a follow on issue.

ustcfy commented:
> The GPU is consistently slower than the CPU is at this. Have you done any profiling so we can understand what is happening?

I haven't done profiling yet, but I will do it next. 👀

ustcfy commented:
> The GPU is consistently slower than the CPU is at this. Have you done any profiling so we can understand what is happening?

It seems that the GPU takes longer to read Parquet files, so I switched to using an in-memory DataFrame.
@thirtiseven pointed out that using `count(bin(a))` can reduce the `C2R` overhead, so I also modified the SQL. Now the end-to-end time is roughly the same.

revans2 reviewed (dismissed):
I ran my own performance numbers and it looks good. 

```
spark.time(spark.range(0,2000000000L,1,32).selectExpr("sum(octet_length(bin(id)))").show())
```

Showed a 20x speedup over 12 CPU cores.

ustcfy commented:
build
